### PR TITLE
feat: Task datasource and fixes

### PR DESCRIPTION
**Pull request by sfc-gh-jcieslak**
## Changes
- Fixed the need to enclose `config` with $$
- Added full datasource support with tests
  - Documentation and examples (v1 candidates; added for resource and datasource)
  - Migration guide (now only for datasource)
  - Extracted schema for regular `in` filtering
  - Describe wasn't added on purpose, because it seems to mirror values from what we get from the SHOW command (can also add it, but for now it doesn't provide any benefits)

## List from previous prs
- [ ] Apply comments https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3170
- [ ] Apply comments from https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3113
  - [x] Cron https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3113#discussion_r1789842978
  - [ ] Resource logic https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3113#discussion_r1789869045
  - [ ] Refactor SDK suspend root logic for tasks (and overall suspend logic in SDK/resource) https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3113#discussion_r1789872814
- [ ] Move some of the logic to SDK (if possible)
  - [ ] https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3170#discussion_r1838382878
- [ ] Refactor task resuming in task resource (most likely with the use of defer) because currently, there may be cases that error can cause tasks to be not resumed.
- [ ] Tests
  - [ ] External changes - https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3113#discussion_r1789903936
  - [ ] Add more complicated DAG structures to show the resource can handle more complex structures
  - [ ] Calling (`as` field) - https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3113#discussion_r1789900590
  - [ ] For showing how the DAG of tasks could be owned by a role other than the one created with Terraform (also with less privileges, only to run the task).
  - [x] Check in one acceptance test why finalizer task in show_output is not set (is that Snowflake or mapping error).
- [x] Data source
- [ ] Examples, documentation, and migration guide
- [ ] Keep manually changed files after regeneration https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3113#discussion_r1802561725
- [x] Make config without $$ escapes needed
- [ ] Support session paramters
- [ ] Analyze non-deterministic test cases
- [ ] Check test tasks_gen_integration_test.go:937 (and see why it's non deterministic).
- [ ] Re-generate and list all the issues with asserts and models

### CODE REVIEW

File: MIGRATION_GUIDE.md [hunk 1]
```diff
@@ -6,6 +6,28 @@ across different versions.
...
+- `starts_with`
+- `root_only`
+- `limit`
+
+New output fields
+- `show_output`
+- `parameters`
+
+Breaking changes:
+- `database` and `schema` are right now under `in` field
```
Comment (sfc-gh-asawicki) [inline MIGRATION_GUIDE.md:27]: As always, please add information what user actions are required to migrate to the new resource/ data source

File: docs/data-sources/tasks.md [hunk 2]
```diff
@@ -2,42 +2,967 @@
 page_title: "snowflake_tasks Data Source - terraform-provider-snowflake"
 subcategory: ""
 description: |-
-  
+  Datasource used to get details of filtered tasks. Filtering is aligned with the current possibilities for SHOW TASKS https://docs.snowflake.com/en/sql-reference/sql/show-tasks query. The results of SHOW and SHOW PARAMETERS IN are encapsulated in one output collection tasks.
 ---
 
-# snowflake_tasks (Data Source)
+!> **V1 release candidate** This data source was reworked and is a release candidate for the V1. We do not expect significant changes in it before the V1. We will welcome any feedback and adjust the data source if needed. Any errors reported will be resolved with a higher priority. We encourage checking this data source out before the V1 release. Please follow the [migration guide](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/MIGRATION_GUIDE.md#v0920--v0930) to use it.
```
Comment (sfc-gh-asawicki) [inline docs/data-sources/tasks.md:8]: v0920--v0930 -> v0980--v0990

File: docs/data-sources/tasks.md [hunk 3]
```diff
@@ -2,42 +2,967 @@
...
 
+# snowflake_tasks (Data Source)
 
+Datasource used to get details of filtered tasks. Filtering is aligned with the current possibilities for [SHOW TASKS](https://docs.snowflake.com/en/sql-reference/sql/show-tasks) query. The results of SHOW and SHOW PARAMETERS IN are encapsulated in one output collection `tasks`.
```
Comment (sfc-gh-asawicki) [inline docs/data-sources/tasks.md:12]: nit (general): we should think of being consistent: data source or datasource (above in v1 release candidate note there is data source and in this paragraph datasource). Let's maybe add it to our task to correct the docs before the v1?

File: docs/resources/task.md [hunk 4]
```diff
@@ -5,6 +5,8 @@ description: |-
   Resource used to manage task objects. For more information, check task documentation https://docs.snowflake.com/en/user-guide/tasks-intro.
 ---
 
+!> **V1 release candidate** This resource was reworked and is a release candidate for the V1. We do not expect significant changes in it before the V1. We will welcome any feedback and adjust the resource if needed. Any errors reported will be resolved with a higher priority. We encourage checking this resource out before the V1 release. Please follow the [migration guide](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/MIGRATION_GUIDE.md#v094x--v0950) to use it.
```
Comment (sfc-gh-asawicki) [inline docs/resources/task.md:8]: v0940--v0950 -> v0980--v0990

File: pkg/sdk/tasks_impl_gen.go [hunk 5]
```diff
@@ -226,6 +228,9 @@ func (r *CreateOrAlterTaskRequest) toOpts() *CreateOrAlterTaskOptions {
 			UserTaskManagedInitialWarehouseSize: r.Warehouse.UserTaskManagedInitialWarehouseSize,
 		}
 	}
+	if r.Config != nil {
+		opts.Config = String(fmt.Sprintf("$$%s$$", *r.Config))
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/tasks_impl_gen.go:232]: should it be added to the migration notesa too? (I mean if someone had dollar signs added by hand previously?)

File: pkg/datasources/common.go [hunk 6]
```diff
@@ -114,6 +145,30 @@ func handleLimitFrom(d *schema.ResourceData, setField **sdk.LimitFrom) {
 	}
 }
 
+func handleIn(d *schema.ResourceData, setField **sdk.In) error {
+	if v, ok := d.GetOk("in"); ok {
+		in := v.([]any)[0].(map[string]any)
+		if v, ok := in["account"]; ok && v.(bool) {
```
Comment (sfc-gh-asawicki) [inline pkg/datasources/common.go:151]: nit: only one of these fields should be set but this is on the schema level, so maybe switch would be better here (or if-elses chain)?

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: MIGRATION_GUIDE.md (hunk 1 above)
Comment (sfc-gh-jcieslak) [reply MIGRATION_GUIDE.md:27]: Done

File: pkg/sdk/tasks_impl_gen.go (hunk 5 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/tasks_impl_gen.go:232]: True, added

File: docs/data-sources/tasks.md (hunk 3 above)
Comment (sfc-gh-jcieslak) [reply docs/data-sources/tasks.md:12]: https://developer.hashicorp.com/terraform/language/data-sources here it's separate, so maybe let's do the same? Changed.

File: pkg/acceptance/bettertestspoc/assert/resourceshowoutputassert/user_show_output_ext.go [hunk 7]
```diff
@@ -1,9 +1,27 @@
 package resourceshowoutputassert
 
 import (
+	"testing"
+
 	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/acceptance/bettertestspoc/assert"
 )
 
+// TaskDatasourceShowOutput is a temporary workaround to have better show output assertions in data source acceptance tests.
+func TaskDatasourceShowOutput(t *testing.T, name string) *TaskShowOutputAssert {
```
Comment (sfc-gh-jmichalak) [inline pkg/acceptance/bettertestspoc/assert/resourceshowoutputassert/user_show_output_ext.go:10]: Why is this added to `user_show_output`?

File: pkg/datasources/tasks.go [hunk 8]
```diff
@@ -2,53 +2,53 @@ package datasources
...
+		Optional:    true,
+		Default:     true,
+		Description: "Runs SHOW PARAMETERS FOR TASK for each user returned by SHOW TASK. The output of describe is saved to the parameters field as a map. By default this value is set to true.",
 	},
-	"schema": {
-		Type:        schema.TypeString,
-		Required:    true,
-		Description: "The schema from which to return the tasks from.",
+	"like":        likeSchema,
+	"in":          inSchema,
```
Comment (sfc-gh-jmichalak) [inline pkg/datasources/tasks.go:23]: Why not extended `in`? I can see that you can filter by apps and app packages as well.

File: MIGRATION_GUIDE.md [hunk 9]
```diff
@@ -6,6 +6,83 @@ across different versions.
 
 > [!TIP]
 > We highly recommend upgrading the versions one by one instead of bulk upgrades.
+ 
+## v0.98.0 ➞ v0.99.0
+
+### snowflake_task resource changes
+
+Breaking changes:
+- `config` field now can be defined without $ signs
```
Comment (sfc-gh-jmichalak) [inline MIGRATION_GUIDE.md:15]: question: `can` or `must`? I mean when users pass with `$`, it won't work, right?

File: MIGRATION_GUIDE.md [hunk 10]
```diff
@@ -6,6 +6,83 @@ across different versions.
...
+
+Before:
+```terraform
+resource "snowflake_task" "config_example" {
+  # ...
+  config = "$$${\"my_config_key\":\"my_config_value\"}$$"
```
Comment (sfc-gh-jmichalak) [inline MIGRATION_GUIDE.md:21]: `$$$` or `$$`?

**Review by sfc-gh-jmichalak — CHANGES_REQUESTED** [review]

File: MIGRATION_GUIDE.md (hunk 9 above)
Comment (sfc-gh-jcieslak) [reply MIGRATION_GUIDE.md]: I didn't notice, but this field was previously not available in the resource, so I'm going to just change it as `new field`. I'll add more in the next pr.

File: MIGRATION_GUIDE.md (hunk 10 above)
Comment (sfc-gh-jcieslak) [reply MIGRATION_GUIDE.md]: It was `$$$` because `${` is treated as string interpolation in HCL and you have to escape it with `$${`

**Review by sfc-gh-jmichalak — APPROVED** [review]

File: pkg/datasources/common.go [hunk 11]
```diff
@@ -148,22 +148,21 @@ func handleLimitFrom(d *schema.ResourceData, setField **sdk.LimitFrom) {
...
-				}
-				*setField = &sdk.In{Schema: schemaId}
+		case okDatabase && databaseValue.(string) != "":
+			*setField = &sdk.In{Database: sdk.NewAccountObjectIdentifier(databaseValue.(string))}
+		case okSchema && schemaValue.(string) != "":
+			schemaId, err := sdk.ParseDatabaseObjectIdentifier(schemaValue.(string))
+			if err != nil {
+				return err
 			}
+			*setField = &sdk.In{Schema: schemaId}
```
Comment (sfc-gh-asawicki) [inline pkg/datasources/common.go:165]: nit (next PR): default return err?

**Review by sfc-gh-asawicki — APPROVED** [review]

File: pkg/datasources/tasks.go [hunk 12]
```diff
@@ -2,53 +2,53 @@ package datasources
...
 var tasksSchema = map[string]*schema.Schema{
-	"database": {
-		Type:        schema.TypeString,
-		Required:    true,
-		Description: "The database from which to return the schemas from.",
+	"with_parameters": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Default:     true,
+		Description: "Runs SHOW PARAMETERS FOR TASK for each user returned by SHOW TASK. The output of describe is saved to the parameters field as a map. By default this value is set to true.",
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/datasources/tasks.go:20]: Addressed in: https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3202

File: pkg/datasources/common.go (hunk 11 above)
Comment (sfc-gh-jcieslak) [reply pkg/datasources/common.go:165]: Addressed in: https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3202
